docs/stuck-ingests.md

```diff
@@ -0,0 +1,130 @@
+# "Stuck" ingests
+
+[Trello card](https://trello.com/c/gdWr7xIo/1138-returns-take-over-two-weeks-to-process-get-stuck)
+
+## The symptoms
+
+- Supplier continues to see "Processing" screen for a submission, even days after submitting it. There is nothing that the supplier can do to resolve this situation.
+- At a data level, this means that the `SubmissionFile` is stuck with `aasm_state == 'processing'`
+
+## Why it happens
+
+1. High RAM usage during `SubmissionIngestionJob` causes the Sidekiq worker to hit its RAM limit and get killed, in the middle of various jobs (the `SubmissionIngestionJob` that caused the error, and any other jobs that are running at the same time, most notably `ManagementChargeCalculationJob`).
+2. Our [Sidekiq config](#we-want-automatic-recovery) means that the killed jobs are then lost and not performed again, even once Sidekiq is restarted, so jobs are never resumed and hence the submission remains stuck.
+
+At the time of writing, the Sidekiq RAM limit on production was [2 GiB](https://github.com/dxw/DataSubmissionServiceAPI/blob/c16cf6b/CF/deploy-app.sh#L108).
+
+### Why high RAM usage?
+
+There are two stages in the ingest process where the RAM usage is very high:
+
+1. Converting the Excel file to CSV using the `in2csv` tool. The memory consumption of this tool increases with the size of the input file (at least, for XLSX files, which is what I tested; Leeky suggested that the structure of an XLS file means that they may be easier to stream from disk).
+2. Reading the CSV rows and validating them, in `Ingest::Loader`. We create an [in-memory array](https://github.com/dxw/DataSubmissionServiceAPI/blob/36252c8598ec44865d4a5eefa18b077e88116c13/app/models/ingest/loader.rb#L62) of `SubmissionEntry` instances, which we validate one-by-one and which we then afterwards persist to the database in a single transaction. This array grows in an unbounded fashion with the size of the input file.
+
+So, we see that this is a problem which is more likely to occur for _large files_.
+
+## When it's happened
```

Comment:
The Trello card references 'large returns'  on 45MB. Do we have any data more on:

- the average file size users attempt, as a bonus I'd also like to know if this size is growing or staying about the same over time
- the size of the larger files we see, what's the upper limit? This is an edge case but if a user is doing something unexpected and uploading 10gb of data for processing, it might be a important to not only prevent processing by rejecting the upload in the first place with a meaningful user error
- out of all monthly submissions, what percentage get stuck? If I were the person picking up this issue I'd be very grateful for clear information on the impact of the issue, and if we don't know, it is still valuable to say so.

Reply:
I've added some information about the distribution of the file sizes that we've seen on production, which hopefully answers these questions!